README.adoc

```diff
@@ -0,0 +1,144 @@
+
+== What you will build
+
+You will build a Vert.x application that exposes and uses `BarmanService`, an Event Bus Service that generates beers and manages customer bills.
+We are going to deploy two verticles: the `BarmanVerticle` that exposes the service and the `DrunkVerticle` that consumes the service
+
+== What you need
+
+* A text editor or IDE
+* Java 8 higher
+* Maven
+
+== Create a project
+
+Here is the content of the `pom.xml` file you should be using:
+
+[source,xml,role="collapsed"]
+.Maven `pom.xml`
+----
+include::pom.xml[]
+----
+
+You must import both `vertx-codegen` with `processor` classifier and `vertx-service-proxy`. We are going to use `vertx-web-client` to load the beer names.
+
+== Design the `BarmanService`
+
+First, we must define our service interface. Behind the hoods, a code generator analyze it and generates the event bus message handler and the proxy class.
+
+This is the `BarmanService` interface:
+[source,java,indent=0,role="collapsed"]
+----
+include::src/main/java/io/vertx/howtos/ebservice/beers/BarmanService.java[]
+----
+<1> Add `@VertxGen` and `@ProxyGen` to trigger code generation
+<2> Define the method that generates a new beer and adds it to the bill of the customer specified. When the beer is ready, It calls the `handler` with the new beer
+<3> Define the method that retrieves the bill.
+<4> Define the method that settles the bill. This is a fire and forget method
+<5> Specify the method that creates a new proxy. `BarmanServiceVertxEBProxy` is the class name of the proxy that the code generator will create for us
+
+There are a couple of rules you must follow while defining the service interface. Look at http://vertx.io/docs/vertx-service-proxy/java/#_restrictions_for_service_interface[Restrictions for service interface] for more info.
+In our interface we are using a couple of primitives and `Beer`, a POJO annotated as `@DataObject`.
+If you want to use `@DataObject`s inside your interface, they must define both a constructor with only a `JsonObject` parameter and a `toJson()` method that returns a `JsonObject`
+
+To trigger the code generation, you must also add in the same package of the interface a `package-info.java` with `@ModuleGen` annotation:
+[source,java,indent=0,role="collapsed"]
+----
+include::src/main/java/io/vertx/howtos/ebservice/beers/package-info.java[]
+----
+
+== Implement the `BarmanService`
+
+Now you can implement the `BarmanService`.
+
+For `giveMeARandomBeer()`:
+[source,java,indent=0,role="collapsed"]
+----
+include::src/main/java/io/vertx/howtos/ebservice/beers/impl/BarmanServiceImpl.java[tags=giveMeARandomBeer]
+----
+<1> Send a request to the https://www.craftbeernamegenerator.com[Craft Beer Name Generator Service]
+<2> Fail the async method if the request to external service failed
+<3> Create a new `Beer`
+<4> Add the beer to the customer bill
+<5> Complete the async method with the generated beer
+
+For `getMyBill()`:
+[source,java,indent=0,role="collapsed"]
```

Comment:
Not sure we want code snippets to be collapsed. For build files yes, but code should not be collapsable. This remark also applies everywhere else.